cmd/trcd/main.go

```diff
@@ -29,31 +35,145 @@ func init() {
 		}
-		defer c.Close()
-
-		go func() {
-			dec := json.NewDecoder(c)
-			dec.DisallowUnknownFields()
-			var msg api.Message
-			if err := dec.Decode(&msg); err != nil {
-				log.Printf("Failed to decode input: %s", err)
-			}
-			log.Printf("Got: %v", msg)
-		}()
-		if _, err := io.Copy(c, os.Stdin); err != nil {
-			log.Printf("Failed to copy input: %s", err)
-			return
+
+	case *tcpSock != "":
+		logger := log.WithField("addr", *tcpSock)
+
+		logger.Info("Listening on TCP socket...")
+		netLst, err = net.Listen("tcp", *tcpSock)
+		if err != nil {
+			logger.WithError(err).Fatal("Failed to listen on TCP socket")
 		}
 	}
+
+	defer netLst.Close()
+
+	closeCh := make(chan struct{})
+
+	go func() {
+		for {
+			sockConn, err := netLst.Accept()
+
+			select {
+			case <-closeCh:
+				if sockConn != nil {
+					err = sockConn.Close()
+					if err != nil {
+						log.WithError(err).Error("Failed to close connection")
+					}
+				}
+				return
+
+			default:
+			}
+
+			if err != nil {
+				log.WithError(err).Error("Failed to accept connection")
+				continue
+			}
+
+			go func() {
+				defer sockConn.Close()
+
+				log.WithField("addr", sockConn.RemoteAddr()).Infof("Connection accepted")
+
+				// state handler of mock TRC, logs all actions
+				setStateHandler := func(msg *api.Message) (*api.Message, error) {
+					log.Infof("Received: %s", msg)
+
+					reply, err := trctest.DefaultStateHandler(msg)
+					log.Debugf("Reply with: %s", reply)
+					return reply, err
+				}
+				// ping handler of mock TRC, logs when ping is received
+				pingHandler := func(msg *api.Message) (*api.Message, error) {
+					log.Debug("Received ping")
+					return trctest.DefaultPingHandler(msg)
+				}
+
+				trcConn := trctest.Connect(sockConn, sockConn,
+					trctest.WithHandler(api.MessageTypeState, setStateHandler),
+					trctest.WithHandler(api.MessageTypePing, pingHandler),
+					trctest.WithHandler(api.MessageTypeHandshake, trctest.DefaultHandshakeHandler),
+				)
+				defer trcConn.Close()
+
+				go func() {
+					for err := range trcConn.Errors() {
+						log.WithError(err).Error("Internal mock-trc error")
+						return
+					}
+				}()
+
+				if err := trcConn.SendHandshake(&api.Handshake{
+					Version: trcapi.DefaultVersion,
+					Token:   "test",
+				}); err != nil {
+					log.WithError(err).Error("Failed to send handshake")
+					return
+				}
+				log.WithField("version", trcapi.DefaultVersion).Debug("Sent handshake")
+
+				if err := trcConn.SendState(apitest.RandomState()); err != nil {
+					log.WithError(err).Error("Failed to send state")
+					return
+				}
+
+				if *silent {
+					<-make(chan int)
+					return
+				}
+
+				for {
+					select {
+					case <-time.After(5*time.Second + time.Millisecond*time.Duration(rand.Intn(3000))):
+						if err := trcConn.SendState(apitest.RandomState()); err != nil {
+							log.WithError(err).Error("Failed to send state")
+							return
+						} else {
```

Comment:
if block ends with a return statement, so drop this else and outdent its block (move short variable declaration to its own line if necessary)